### PR TITLE
initial implementation of scale type

### PR DESCRIPTION
**Pull request by orndorffgrant**
If this PR fixes a bug, you _must_ add test cases representative of the bug.

Please refer to [Tettra](https://app.tettra.co/teams/amida/pages/amida-pull-request-and-code-review-guide) for PR review guidelines.

#### What does this PR do?
This adds the "scale" question type, which is allows min and max limits on a number input. This was adapted from a similar feature added to survey service. See https://github.com/amida-tech/amida-survey-microservice/pull/38

#### Related JIRA tickets:
RR-756

#### How should this be manually tested?
Create/get/export/import scale questions and answer them.

1. Scale questions are specified using type scale and property meta: { scaleLimits: {min: , max: }}.
2. Min and max are optional but at least one of them should be present.
3. When both present max needs to be greater or equal to min.
4. Answers to scale questions are specified using numberValue which accepts float.
5. Answers will be validated when posted according to min and max.

#### Screenshots (if appropriate):

### CODE REVIEW

**Comment by orndorffgrant** [comment]
Pretty sure the swagger.json additions are unnecessary/wrong but wanted to get this up for review.

**Comment by orndorffgrant** [comment]
This is currently its own question type and accepts float values. That was the feature added to survey service so I just did the same thing.

Now that I understand the code base better I could change it to just be extra validations on integer or float questions iff there are limits in the meta field.

**Comment by unprofessional** [comment]
There are a couple of eslint errors:
```
/lib/json-schema.js 607:13  error  Duplicate key 'meta'  no-dupe-keys
/test/util/generator/answerer.js97:10  error  Expected 'this' to be used by class method 'scale'  class-methods-use-this
```
Use `grunt eslint` to see them.

**Comment by unprofessional** [comment]
Also one test is failing for me:
``
(migration.spec.js 137)
1) migration spec compare question_type records:

      AssertionError: expected [ Array(17) ] to deeply equal [ Array(18) ]
      + expected - actual

         {
           "name": "pounds"
         }
         {
      +    "name": "scale"
      +  }
      +  {
           "name": "text"
         }
         {
           "name": "year"
```
It looks like the `question_type` migration-model disagrees with the canonical model. This could be be because I neglected to do something before running the tests, but these tests are self-contained (i.e. this one runs a separate `recregmigrations` database instance).

@au2 , would Grant need to perform a migration for the updated value-in-array `scale` in the canonical `question_type` model? Or am I way off base here?

**Comment by au2** [comment]
yes, we need a migration for the added type and columns

**Comment by orndorffgrant** [comment]
I removed the scaleLimits defs from swagger.json because they are being sent inside the meta field.

I don't fully understand the migration tests. The change I made (adding 'scale' to the list of question types) lets the test pass but do I need to do anything else?

**Comment by au2** [comment]
If there are existing customers anything that changes the database needs a migration script that can be applied when customers update. Unlike services this project has an existing customer. We use sequelize-cli for migration. Take a look at the migrations directory to see some up the previous migration scripts.

**Comment by au2** [comment]
I do not think I used meta for scale questions in survey service. So I am not sure what you are referring to. I will try to take a look at the code today sometime.

**Comment by orndorffgrant** [comment]
Thank you for the explanation @au2 !

I changed the implementation - partially because Alan said something about using the meta field, largely because I knew I'd learn more if I did something beyond copy/pasting.

Which means this pull request is not the exactly the same even though it serves the same purpose (and is why I called the swagger.json additions unnecessary).

In this implementation, the client includes `scaleLimits { ... }` inside `meta` in it's request. Then the backend stores `scaleLimits` inside `meta` in the db - so there is not a new `parameter` column.

Though, if we want this backend's api to be the same as survey-service then I should change things back to how you did them.

**Comment by au2** [comment]
Fair. But if you are going to use Meta you do not need the new question
type either. You can just tag that along to meta as well with a number or
integer question type. Either way.

**Comment by orndorffgrant** [comment]
To keep future migration to use survey service  easier, I changed it back to be the same as https://github.com/amida-tech/amida-survey-microservice/pull/38.

I added a migration that just adds the parameter column to the question table with default value null.

File: migration/models/question-type.model.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ module.exports = function Table(sequelize, DataTypes) {
                     const names = [
                         'text', 'choice', 'choices', 'bool', 'integer', 'float',
                         'zip', 'date', 'pounds', 'year', 'month', 'day',
-                        'feet-inches', 'blood-pressure', 'choice-ref',
+                        'feet-inches', 'blood-pressure', 'choice-ref', 'scale',
```
Comment (au2) [inline migration/models/question-type.model.js]: You do not change files in the migration/models directory. You need to insert scale in your migration script.

File: migration/migrations/20180206155614-scale-question-type.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.addColumn('question', 'parameter', {
+        type: Sequelize.TEXT,
+        allowNull: true,
+        defaultValue: null
+    });
+  },
+};
```
Comment (au2) [inline migration/migrations/20180206155614-scale-question-type.js]: Run eslint for this file. And insert 'scale' to question types here.

**Comment by au2** [comment]
We also need to update patch for this. Take a look at the question_patch tests and I think you can add new test for scale limits and go from there for implementation.

**Comment by au2** [comment]
Also please merge develop. Looks like my merges introduced conflicts with your development.

**Comment by orndorffgrant** [comment]
Will do. Thank you for the help! I'll reach out if I have questions

**Comment by orndorffgrant** [comment]
I think I got everything you mentioned @au2 
Specifically make sure I did the migration correctly, and that the patch tests are sufficient.

File: models/dao/question.dao.js [hunk 3]
```diff
@@ -623,6 +657,16 @@ module.exports = class QuestionDAO extends Translatable {
                     }
                     return null;
                 });
+                const { scaleLimits } = patch;
+                if (scaleLimits && question.scaleLimits && question.type === 'scale') {
+                    if (!_.isEqual(question.scaleLimits, scaleLimits)) {
+                        const combined = _.cloneDeep(question.scaleLimits);
+                        Object.assign(combined, scaleLimits);
```
Comment (au2) [inline models/dao/question.dao.js]: I not sure what you are trying to accomplish here. Can't you just do parameterJsonToString({ scaleLimits }) without doing any clones or assigns?

**Comment by au2** [comment]
Sorry for piecemeal comments but I think you will need to update surveyQuestionPatch object in swagger and add updates to survey_patch.model and survey_patch.integration tests. Take a look at those test, they are a but overengineered so let me know if you need any help understanding them. You need to add a scale type question to some of the surveys.

File: models/dao/question.dao.js (hunk 3 above)
Comment (orndorffgrant) [reply models/dao/question.dao.js]: We don't allow nulls for min or max in patches? In that case, then yes this is unnecessary (as well as similar merging in the test patching).
That does limit the capabilities though: if a question had only a min on creation, then we wanted to update the min, we'd have to give it a max as well. Is that okay?

**Comment by au2** [comment]
Yes, it should be OK. Unfortunately Swagger 2.0 does not support null's for
fields. Open API 3.0 (aka Swagger 3.0) will support nulls but the library
we are using (swagger-tools) is not there yet.

**Comment by orndorffgrant** [comment]
@au2 I updated the surveyQuestionPatch object in swagger. 

It looks like the survey_patch tests generate questions using question-generator.js which generates scale type questions every X questions, so scale questions are being included in the test surveys. I verified that scale questions are present via console.log and grep. Do I need to do anything else?

**Comment by au2** [comment]
Was it failing before you made changes to surveyQuestionPatch?

**Comment by orndorffgrant** [comment]
Hmmm I just removed scaleLimits from surveyQuestionPatch and the tests still passed. What does that mean?

**Comment by au2** [comment]
That means you are not patching any of the scale questions in the test. You
need to patch at least have one case where you patch a scale type questions
(change min, max or both).

**Comment by orndorffgrant** [comment]
@au2 last commit should do it. Check additions to test/fixtures/conditional-session/patch/patch-setup.js and test/util/generator/survey-patch-generator.js

**Comment by au2** [comment]
1) Please run grunt or grunt eslint. Looks like you did not run eslint.
2) I am not sure how you found the scale question. If it is by trial and error I would update setup  to make a question of type scale explicitly.

**Comment by orndorffgrant** [comment]
Thank you! I was unaware of to how to explicitly insert a question but you pointed me in the right direction. 

I think this is finally good to go @au2

**Pull request merged**